lib/rules/no-extra-parens.js

```diff
@@ -602,17 +653,79 @@ module.exports = {
             },
 
             ForStatement(node) {
-                if (node.init && hasExcessParens(node.init)) {
-                    report(node.init);
-                }
-
                 if (node.test && hasExcessParens(node.test) && !isCondAssignException(node)) {
                     report(node.test);
                 }
 
                 if (node.update && hasExcessParens(node.update)) {
                     report(node.update);
                 }
+
+                if (node.init) {
+                    forStatementInitInfo = {
+                        upper: forStatementInitInfo,
+                        inExpressionNodes: [],
+                        reports: []
+                    };
+
+                    if (hasExcessParens(node.init)) {
+                        report(node.init);
+                    }
+                }
+            },
+
+            "ForStatement > *.init:exit"(node) {
+
+                /*
+                 * Removing parentheses around `in` expressions might change semantics and cause errors.
+                 *
+                 * For example, this valid for loop:
+                 *      for (let a = (b in c); ;);
+                 * after removing parentheses would be treated as an invalid for-in loop:
+                 *      for (let a = b in c; ;);
+                 */
+
+                let { reports } = forStatementInitInfo;
+
+                if (reports.length) {
+
+                    const { inExpressionNodes } = forStatementInitInfo;
+
+                    inExpressionNodes.forEach(inExpressionNode => {
+                        for (const pathNode of pathToAncestor(inExpressionNode, node).reverse()) {
+                            if (isParenthesised(pathNode)) {
```

Comment:
I guess any of the following nodes should `break` this loop because wrapping by any of `()`, `[]`, `{}`, and `?:` allows `in` operators.

- Wrapping by `()`
    - `ArrowFunctionExpression` if the child is in `params`.
    - `CallExpression` if the child is in `arguments`.
    - `FunctionExpression` if the child is in `params`.
    - `NewExpression` if the child is in `arguments`.
- Wrapping by `[]`
    - `ArrayExpression`
    - `ArrayPattern`
    - `MemberExpression` if the child is `property` and `computed` is `true`.
- Wrapping by `{}`
    - `BlockStatement`
    - `ObjectExpression`
    - `ObjectPattern`
    - `TemplateLiteral`
- Wrapping by `?:`
    - `ConditionalExpression` if the child is `consequent`.

Reply:
I was aware of this, but eventually decided it might be better not to implement this check, and left a comment at the end of the loop:

```js
 /*
  * Other enclosing punctuators such as {} are omitted from this check,
  * in order to avoid dependencies with other rules.
  *
  * For example, these parentheses might be safe for removal:
  *      for (let a = b => { return (b in c) }; ;);
  * However, arrow-body-style could remove braces as well, and produce an invalid for-in loop:
  *      for (let a = b => b in c; ;);
  */
```

Though now I'm not sure about this particular example because the ranges overlap, perhaps the fixing algorithm would have already removed the braces in the previous iteration (while blocking the parens fix in that iteration).

Also, this additional check might be too much code for an edge case with not-so-terrible consequences (one pair of unnecessary parens will stay) of an already edge case ('in' in a for-loop).

Of course, if this additional check can't produce subsequent errors, and you think it's worth of implementing, I'll do it.

Reply:
ESLint doesn't apply two fixes which have overlapped range each other at the same time. So I think this will not conflict with `arrow-body-style` rule.

Reply:
Ok, I'll do it by the list you provided